sticker/src/tensorflow/dataset.rs

```diff
@@ -209,3 +231,61 @@ where
         None
     }
 }
+
+/// An Iterator adapter performing local shuffling.
+///
+/// Fills a buffer with size `buffer_size` on the first call. Subsequent
+/// calls swap the next incoming item with a random element from the
+/// buffer and return the random element.
+pub struct Shuffled<I> {
+    inner: I,
+    buffer: Vec<Result<Sentence, Error>>,
```

Comment:
Maybe this should be `Vec<Sentence>` so that I/O errors fail as early as possible?

Reply:
The `Item` type of `SentenceIter` is `Result<Sentence, Error>>`. If we want to fail as early as possible, we'd need to change that type.

Reply:
The iterator needs that type, because iterators are lazy. But the buffering is a greedy, so it could unwrap sentences and if it hits an `Err` return `Some(Err(...))` immediately.

Reply:
Not sure if I'm understanding this. 

If `self.buffer` is a `Vec<Sentence>`, we need to unwrap the incoming sentences, check for Err, return if there is one and then re-wrap the non-Err values in `Ok` to comply with the type of `SentenceIter`. 

```rust
match sent {
    Ok(sent) => buffer.push(sent),
    e => return Some(e)
}

sent = get_random_sent();

Some(Ok(sent))
```

Or did you mean something like:

```rust
            while let Some(sent) = self.inner.next() {
                if sent.is_err() {
                    return sent
                }
                self.buffer.push(sent);
                if self.buffer.len() == self.buffer_size {
                    break;
                }
            }
```
In both cases, we eliminated the Err variant of the type of `SentenceIter`.

Reply:
The former, since the latter would still require buffer to be `Vec<Result<...>>`. Rewrapping in `Ok` should be very cheap. Since there is a swap (which is a copy) anyway, it will probably be swapped in-place on the stack.